readme.md
# Torpedo

Torpedo is a proof of concept Tor client implementation of the Tor procotol, written in pure Rust

**Note**

This product is produced independently from the Tor® anonymity software and carries no guarantee from [The Tor Project](https://www.torproject.org/) about quality, suitability or anything else.



changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]
### To Do

#### Torpedo

- Only fetch a new consensus/router data if expired. Otherwise try to use cached data
- Don't keep allocating memory (via Vec) try to use one global Vec and reuse it for multiple cells
- Finish `TorCell::try_from_stream` which keeps calling `TorCell::from_stream` until it succeeds
- Finish the code that ensures that invalid cells are discarded correctly

#### torserde_macros
- Clean up macro code, create functions for repeated code

#### Torserde
- Rename `NLengthVector` to something less confusing

### Unfinished Ideas
None

## [0.1.6] - 2021-07-02
### Added
- Support for serialising the discriminant as the type specified by the `repr` attribute
  - `process_attributes` added to torserde_macros to obtain any required attributes form the deriving object
    - At the moment this only extracts the `repr` attribute, but may be expanded to any other relevant attributes
  - This means that if, say, `#[repr(u16)]` is specified then it will serialise the discriminant as a u16
  - Tests have been added for non-u8 repr
- `Command::is_var_command` determines whether a command (of type `u128` not `Command`) is variable length or not 
- `torserde::ErrorKind::BadDiscriminant` now returned when an unknown command byte is deserialised from an enum
- Added implementation of `Torserde` for `u64`

## [0.1.5] - 2021-07-01
### Added
- `torserde::ErrorKind::InvalidRelayLength` error added to show when a serialised cell is not the correct length of 509 bytes
- misc.rs for particular snippets belonging to no specific branch of Torpedo
- `misc::UnpackedCell` added to help with the fast serialization and deserialization of `Relay`

### Removed
- 'Clone' trait from various types

### Fixed
- `NullReader` renamed to `NullStream` and the `write` function now returns the number of bytes read

## [0.1.4] - 2021-06-16
### Added
- We now flush padding for variable length cells too
- `torserde::ErrorKind` added to encapsulate Torserde-specific errors (like bad digest, invalid command, etc.) and generic errors from `std::io` and `bincode`  
- Torserde trait now returns `torserde::Result` for serialize and deserialize which uses the `torserde::ErrorKind` enum
- The return value from serialisation (i.e. the running total number of bytes serialised) now includes the discriminant for enums
- Support for `torserde::Result` in macros and in `CellCrypto`

## [0.1.3] - 2021-05-26
### Added
- Python script created that converts mirror and directory list into Rust code
- Relay cells implemented
  - Relay cells contain an `Encrypted` struct which mark the payload as ready for decryption and verification
  - The payload of a decrypted cell is essentially an Option, as cells that should have a payload but have a length of zero are represented as None
  - Cellcrypto modified to work with Cells and Encrypted structs instead of buffers
  - Relay cells are padded with random bytes
  - By default, the command and payload are not contiguous for relay cells. We join them together to ensure they can be serialised/deserialised with Torserde
  - Padding is stored in `RelayCell` so it can be used to calculate the rolling digest
- SendMePayload added
- We use the `ring` CSPRNG to generate random padding 
- Various warnings cleared

### Removed
- RelayCommand enum removed as no longer needed

## [0.1.2] - 2021-05-26
### Added
- CryptoCell struct which handles calculation of forward and backward encryption/decryption and digests for relay nodes
- Demonstration of communication down to relay begin and relay connected level.

## [0.1.1] - 2021-05-12
### Added
- The framework for test on Torserde macros and the Torpedo lib
- Removed `payload_length` argument from `bin_deserialise_from` since payloads now include the length at the beginning
- `VersionsVector` updated to use the length (in bytes) supplied by the cell

### Changed
- `serialised_length` calculation for `[0u8; N]` types has been optimised

## [0.1.0] - 2021-05-05
### Added
- Initial commit
- Created `Torserde` trait exposing functions that serialise/deserialise according to Tor specification
- Implemented `Torserde` for various types
- Created tests for `Torserde`  
- Created `Torserde` derive macro to derive for structs and enums

src/macro_tests.rs


#[cfg(test)]
mod macro_tests {

    use torserde_macros::Torserde;
    use torserde::{TorSerde, NLengthVector};
    use std::net::Ipv4Addr;

    #[derive(Torserde, PartialEq, Eq, Debug)]
    struct SimpleStruct {
        x: u32,
        y: u32,
    }

    #[derive(Torserde, PartialEq, Eq, Debug)]
    #[repr(u8)]
    enum SimpleEnum {
        First = 10,
        Second(u32, Ipv4Addr) = 11,
        Third(u128, String) = 12,
    }

    #[derive(Torserde, PartialEq, Eq, Debug)]
    #[repr(u64)]
    enum ReprEnum {
        R1 = 400,
        R2(u16) = 500,
        R3{test: String} = 6000,
    }

    #[derive(Torserde, PartialEq, Eq, Debug)]
    struct InnerList {
        command: u8,
        list: NLengthVector<SimpleStruct, 1>,
    }

    #[derive(Torserde, PartialEq, Eq, Debug)]
    struct OuterList {
        command: u8,
        list: NLengthVector<InnerList, 1>,
    }

    #[test]

    fn repr_test() {
        let mut buff = vec![];

        let en = ReprEnum::R2(5600);

        en.bin_serialise_into(& mut buff);

        assert_eq!(buff, [0, 0, 0, 0, 0, 0, 1, 244, 21, 224]);

        let buff = [0u8, 0, 0, 0, 0, 0, 1, 144];

        let en = ReprEnum::bin_deserialise_from(buff.as_ref()).unwrap();

        assert_eq!(en, ReprEnum::R1);

    }

    #[test]
    fn nested_list() {
        let buff = vec![
            14,
            2,
                0,
                1,
                    0, 0, 0, 1,
                    0, 0, 0, 2,
                12,
                2,
                    0, 0, 0, 3,
                    0, 0, 0, 4,
                    0, 0, 0, 5,
                    0, 0, 0, 6];

        let outer = OuterList::bin_deserialise_from(buff.as_slice()).unwrap();

        println!("outer: {:?}", outer);
    }

    #[test]
    fn simple_struct_test() {
        let mut buf = Vec::new();

        let simple_struct = SimpleStruct{ x: 0xf2795e6d, y: 0x4e9237a7 };

        simple_struct.bin_serialise_into(& mut buf).unwrap();

        assert_eq!(buf, vec![0xf2, 0x79, 0x5e, 0x6d, 0x4e, 0x92, 0x37, 0xa7]);

        let d_simple_struct = SimpleStruct::bin_deserialise_from(buf.as_slice()).unwrap();

        assert_eq!(simple_struct, d_simple_struct);



    }

}



src/torpedo_tests.rs


#[cfg(test)]
mod torpedo_tests {
    use native_tls::{TlsConnector, Protocol};
    use std::net::{TcpStream, IpAddr, Ipv4Addr};
    use crate::cells::{TorCell, Command, RelayCell, Encrypted};
    use torserde::{NLengthVector, VersionsVector};
    use torserde::TorSerde;
    use chrono::Local;
    use std::str::FromStr;

    use crate::custom_crypto::kdf_tor;
    use crate::cellcrypto::CellCrypto;
    use std::convert::TryInto;
    use crate::misc::UnpackedCell;

    use rand::Rng;
    use std::io::{Read, Write};

    #[test]
    fn test_cells_coms() {
        let sample = [("148.251.23.114", "443"),
            ("212.129.32.142", "9001"),
            ("185.220.100.252", "443"),
            ("51.159.57.128", "9001"),
            ("148.251.67.153", "9001"),
            ("87.236.195.253", "143"),
            ("95.216.1.251", "8443"),
            ("45.153.160.135", "9001"),
            ("62.210.205.228", "443"),
            ("45.153.160.138", "9001")];

        let mut generator = rand::thread_rng();

        let index = generator.gen_range(0..sample.len());

        let ip = sample.get(index).unwrap();

        let connector = TlsConnector::builder()
            .danger_accept_invalid_hostnames(true)
            .danger_accept_invalid_certs(true)
            .min_protocol_version(Some(Protocol::Tlsv12))
            .build().unwrap();

        let stream = TcpStream::connect(format!("{}:{}", ip.0, ip.1)).unwrap();
        let mut stream = connector.connect("", stream).unwrap();

        TorCell::new(0, Command::Versions {version_list: VersionsVector::from(vec![3, 4]) }).into_stream(& mut stream, 3).unwrap();

        let versions = TorCell::from_stream(& mut stream, 3).unwrap();

        println!("{:?}", versions);

        let certs = TorCell::from_stream(& mut stream, 4).unwrap();

        println!("{:?}", certs);

        let auth = TorCell::from_stream(& mut stream, 4).unwrap();

        println!("{:?}", auth);

        let netinfo = TorCell::from_stream(& mut stream, 4).unwrap();

        println!("{:?}", netinfo);

        TorCell::new(0, Command::NetInfo {
            timestamp: Local::now(),
            other_ip: IpAddr::V4(Ipv4Addr::from_str(ip.0).unwrap()),
            this_ips: NLengthVector::from(vec![IpAddr::V4(Ipv4Addr::from_str("0.0.0.0").unwrap())]) }).into_stream(& mut stream, 4).unwrap();

        let half_handshake = [3u8; 20];

        TorCell::new(0x80000001, Command::CreateFast {onion_skin: half_handshake}).into_stream(& mut stream, 4).unwrap();

        let created_fast = TorCell::from_stream(& mut stream, 4).unwrap();

        println!("{:?}", created_fast);

        if let Command::CreatedFast{ handshake_data } = created_fast.get_command() {

            //Generate and obtain shared secrets

            let mut shared_secret = Vec::from([3u8; 20]);

            shared_secret.extend_from_slice(&handshake_data[0..20]);

            println!("secret: {:?}", shared_secret);

            let materials = kdf_tor(&shared_secret);

            let mut cell_crypto = CellCrypto::from(&materials[20..92].try_into().unwrap());

            //Send cells

            let mut relay_begin_dir = RelayCell::new(1, crate::cells::Relay::BeginDir);

            cell_crypto.set_forward_digest(& mut relay_begin_dir).unwrap();

            let encrypted_begin = cell_crypto.encrypt(relay_begin_dir).unwrap();

            let relay_begin = TorCell::new(0x80000001, Command::Relay { contents: encrypted_begin });

            relay_begin.into_stream(& mut stream, 4).unwrap();

            //Read cell

            let begun = TorCell::from_stream(& mut stream, 4).unwrap();

            println!("cell data: {:?}", begun);

            if let crate::cells::Command::Relay { contents } = begun.get_command() {
                let mut contents = cell_crypto.decrypt(contents).unwrap();

                println!("decrypted: {:?}", contents);

                let digest_result = cell_crypto.verify_backward_digest(& mut contents);



                println!("Verifyt: {:?}", digest_result);

                println!("Relay: {:?}", contents.get_payload());
            }


        }




    }

    #[test]
    fn test_kdf_tor() {
        use crate::custom_crypto::kdf_tor;

        let secret = [120u8, 177, 253, 48, 253, 132, 68, 100, 0, 13, 186, 177, 249, 202, 249, 125, 25, 170, 184, 108, 70, 232, 121, 233, 100, 6, 158, 233, 138, 146, 107, 137, 220, 35, 116, 88, 176, 221, 248, 163, ];

        kdf_tor(&secret);

    }

    #[test]
    fn test_cell_crypto() {



        let key_materials: [u8; 72] = [44, 102, 210, 192, 76, 172, 8, 173, 133, 217, 207, 18, 129, 148, 34, 186, 219, 97, 121, 226, 167, 46, 22, 191, 124, 86, 64, 116, 90, 33, 94, 66, 220, 175, 77, 207, 126, 245, 190, 175, 93, 187, 116, 16, 2, 169, 37, 146, 211, 13, 242, 210, 79, 237, 174, 147, 43, 8, 58, 20, 144, 50, 192, 237, 0, 68, 185, 81, 204, 103, 69, 93];

        let mut crypto_cell = CellCrypto::from(&key_materials);

        let mut data = [0u8; 509];

        let mut cell = RelayCell::new(1, crate::cells::Relay::BeginDir);

        cell.bin_serialise_into(data.as_mut()).unwrap();

        println!("relay testing: {:?}", data);

        crypto_cell.set_forward_digest(& mut cell).unwrap();
        println!("digest {}", cell.get_digest());

        cell.bin_serialise_into(data.as_mut()).unwrap();

        println!("WITH DIGEST: {:?}", data);

        let encrypted = crypto_cell.encrypt(cell).unwrap();

        println!("ENCRYPTED: {:?}", encrypted.0);

    }

    #[test]
    fn second_cellcrypto() {
        let key_materials: [u8; 72] = [44, 102, 210, 192, 76, 172, 8, 173, 133, 217, 207, 18, 129, 148, 34, 186, 219, 97, 121, 226, 167, 46, 22, 191, 124, 86, 64, 116, 90, 33, 94, 66, 220, 175, 77, 207, 126, 245, 190, 175, 93, 187, 116, 16, 2, 169, 37, 146, 211, 13, 242, 210, 79, 237, 174, 147, 43, 8, 58, 20, 144, 50, 192, 237, 0, 68, 185, 81, 204, 103, 69, 93];

        let mut crypto_cell = CellCrypto::from(&key_materials);

        let encrypted = Encrypted([149, 240, 30, 56, 185, 14, 138, 9, 191, 248, 201, 17, 121, 247, 106, 134, 21, 106, 132, 107, 145, 150, 202, 152, 104, 51, 100, 164, 58, 84, 121, 103, 41, 97, 220, 17, 244, 107, 131, 1, 254, 116, 82, 149, 13, 62, 29, 72, 64, 192, 49, 75, 245, 161, 35, 27, 221, 12, 41, 145, 248, 59, 156, 176, 181, 70, 230, 152, 242, 95, 105, 154, 67, 36, 160, 107, 187, 64, 124, 76, 229, 195, 142, 25, 133, 77, 75, 171, 199, 108, 247, 226, 119, 226, 214, 169, 99, 211, 131, 36, 86, 234, 146, 53, 241, 139, 145, 251, 153, 198, 15, 255, 217, 210, 2, 105, 246, 83, 226, 30, 177, 148, 2, 228, 208, 193, 131, 199, 225, 239, 176, 80, 113, 120, 122, 51, 174, 23, 38, 159, 83, 161, 13, 102, 70, 30, 189, 56, 71, 58, 202, 217, 167, 97, 172, 145, 15, 166, 218, 190, 90, 125, 134, 171, 153, 39, 165, 44, 206, 226, 141, 147, 234, 55, 237, 103, 250, 160, 60, 252, 130, 51, 92, 146, 208, 181, 42, 121, 122, 133, 189, 60, 231, 152, 237, 224, 55, 210, 223, 238, 63, 45, 94, 98, 248, 237, 12, 174, 149, 91, 133, 125, 45, 55, 23, 46, 81, 213, 181, 228, 70, 54, 222, 152, 214, 180, 50, 192, 162, 240, 22, 241, 242, 118, 16, 140, 102, 80, 204, 44, 220, 205, 56, 111, 216, 83, 7, 217, 88, 119, 194, 173, 30, 9, 193, 74, 68, 213, 237, 224, 201, 202, 201, 80, 52, 69, 9, 118, 176, 105, 24, 227, 156, 206, 114, 165, 46, 200, 120, 30, 152, 71, 64, 2, 60, 65, 68, 223, 127, 248, 131, 7, 137, 0, 61, 76, 156, 224, 91, 137, 108, 87, 13, 203, 14, 143, 205, 68, 210, 176, 70, 30, 45, 9, 235, 112, 58, 41, 145, 204, 65, 86, 228, 148, 94, 234, 121, 73, 170, 161, 141, 162, 13, 57, 38, 61, 232, 80, 224, 103, 103, 132, 165, 62, 53, 116, 197, 210, 195, 116, 141, 125, 57, 231, 107, 216, 133, 255, 28, 38, 172, 7, 150, 200, 66, 120, 158, 29, 199, 25, 107, 134, 79, 77, 108, 230, 138, 48, 249, 120, 59, 245, 209, 179, 8, 180, 33, 93, 54, 99, 27, 180, 3, 255, 246, 248, 230, 208, 138, 128, 133, 227, 252, 87, 141, 124, 141, 77, 57, 186, 72, 61, 143, 170, 78, 77, 37, 45, 43, 170, 8, 160, 126, 52, 133, 175, 12, 105, 236, 96, 38, 20, 20, 57, 83, 137, 9, 58, 180, 183, 113, 255, 251, 250, 180, 152, 130, 255, 81, 235, 99, 217, 129, 17, 185, 193, 124, 49, 219, 219, 111, 80, 195, 22, 219, 137, 122, 215, 251, 63, 253, 81, 35, 110, 122, 3, 110, 206, 19, 185, 39, 89, 42, 255, 226, 171, 63, 13, 16, 202, 109, 57, 15, 179, 175, 133, 101, 33, 237, 6, 24, 13, 200, 180, 35, 228, 56, 194, 123]);

        let mut cell = crypto_cell.decrypt(&encrypted).unwrap();

        println!("Decrypted: {:?}", cell);

        println!("digested: {:?}", crypto_cell.verify_backward_digest(& mut cell));


    }

    #[test]
    fn test_mirror_dirs() {
    }

    #[test]
    fn test_read_unpacked() {

        let mut cell = UnpackedCell::new(40, Some(vec![23, 6, 54, 23, 54, 6, 8, 43, 2, 45, 56, 42, 28, 5, 2, 13, 85]));

        let bytes = vec![40, 23, 6, 54, 23, 54, 6, 8, 43, 2, 45, 56, 42, 28, 5, 2, 13, 85];

        let mut ser = Vec::new();

        cell.read_to_end(& mut ser).unwrap();

        assert_eq!(bytes, ser);

    }

    #[test]
    fn test_write_unpacked() {

        let mut cell = UnpackedCell::default();

        let bytes = [17, 87, 23, 72, 33, 38, 145, 234];

        cell.write_all(&bytes).unwrap();

        let bytes = [34, 62, 76, 129];

        cell.write_all(&bytes).unwrap();

        assert_eq!(cell.command(), 17);

        assert_eq!(cell.data(), vec![87, 23, 72, 33, 38, 145, 234, 34, 62, 76, 129]);
    }

}

src/misc.rs
use std::io::{Write, Read};
use std::io::Result;

/// A simple write stream that consumes bytes
pub struct NullStream;

impl Write for NullStream {
    fn write(&mut self, _buf: &[u8]) -> std::io::Result<usize> {
        std::io::Result::Ok(_buf.len())
    }

    fn flush(&mut self) -> std::io::Result<()> {
        std::io::Result::Ok(())
    }
}

/// A stream that serialises and deserialises a command/data pair
pub struct UnpackedCell {
    command: u8,
    data: Option<Vec<u8>>,
    written_command: bool,
    bytes_read: u64,
}

impl UnpackedCell {
    pub fn new(command: u8, data: Option<Vec<u8>>) -> Self {
        Self {
            command,
            data,
            written_command: false,
            bytes_read: 0,
        }
    }

    pub fn command(&self) -> u8 {
        self.command
    }

    pub fn data_length(&self) -> usize {
        self.data.as_ref().unwrap().len()
    }

    pub fn data(self) -> Vec<u8> {
        self.data.unwrap()
    }
}

impl Default for UnpackedCell {
    fn default() -> Self {
        Self::new(0, None)
    }
}

impl Write for UnpackedCell {
    fn write(&mut self, mut buf: &[u8]) -> Result<usize> {

        let mut bytes_written = 0;

        if self.data.is_none() {
            self.data = Some(Vec::with_capacity(509)); //Relay cells are max 509 bytes
        }

        if !self.written_command {
            self.command = *buf.get(0).unwrap();
            buf = &buf[1..];
            bytes_written += 1;
            self.written_command = true;
        }

        let vec_ref = self.data.as_mut().unwrap();

        vec_ref.extend_from_slice(buf);

        bytes_written += buf.len();

        Ok(bytes_written)
    }

    fn flush(&mut self) -> Result<()> {
        Result::Ok(())
    }
}

impl Read for UnpackedCell {
    fn read(&mut self, mut buf: &mut [u8]) -> Result<usize> {

        let mut bytes_read = 0;

        if self.data.is_none() {
            return Err(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "Unpacked cell contains no data"));
        }

        if self.bytes_read == 0 {
            let first_element = buf.get_mut(0).unwrap();
            *first_element = self.command;
            buf = & mut buf[1..];
            bytes_read += 1;
            self.bytes_read += 1;
            self.written_command = true;
        }

        let start = self.bytes_read as usize - 1;

        let vec_ref = self.data.as_ref().unwrap();

        let write_bytes = if start + buf.len() > vec_ref.len() { vec_ref.len() - start } else { buf.len() };

        buf.write_all(&vec_ref[start..start+write_bytes]).unwrap();

        bytes_read += write_bytes;

        self.bytes_read += write_bytes as u64;

        println!("read: {}", self.bytes_read);

        Ok(bytes_read)
    }
}


src/cellcrypto.rs

use sha1::Sha1;
use sha1::Digest;
use std::io::Write;

use ctr::cipher::{NewCipher, StreamCipher};
use std::convert::TryInto;
use crate::cells::{RelayCell, Encrypted};
use torserde::TorSerde;

type Aes128Ctr = ctr::Ctr128BE<aes::Aes128>;

//An object to handle the forward encryption, backward decryption and verification of relay cells
pub struct CellCrypto {
    forward_digest: Sha1,
    backward_digest: Sha1,
    forward_encryptor: Aes128Ctr,
    backward_decryptor: Aes128Ctr,
}

impl From<&[u8; 72]> for CellCrypto {
    fn from(key_materials: &[u8; 72]) -> Self {
        Self::new(
            (&key_materials[0..20]).try_into().unwrap(),
            (&key_materials[20..40]).try_into().unwrap(),
            (&key_materials[40..56]).try_into().unwrap(),
            (&key_materials[56..72]).try_into().unwrap())
    }
}

impl CellCrypto {
    pub fn new(f_digest: &[u8; 20], b_digest: &[u8; 20], f_key: &[u8; 16], b_key: &[u8; 16]) -> Self {
        let mut forward_digest = Sha1::new();
        let mut backward_digest = Sha1::new();

        forward_digest.write(f_digest).unwrap();
        backward_digest.write(b_digest).unwrap();

        //Tor uses an inialisation vector of all zeros in its SHA1 hashes
        let iv = 0u128.to_be_bytes();

        let forward_encryptor = Aes128Ctr::new(f_key.into(), iv.as_ref().into());
        let backward_decryptor = Aes128Ctr::new(b_key.into(), iv.as_ref().into());

        println!("fkey: {:?}", f_key);

        Self {
            forward_digest,
            backward_digest,
            forward_encryptor,
            backward_decryptor,
        }

    }

    pub fn set_forward_digest(& mut self, relay: & mut RelayCell) -> torserde::Result<()> {

        relay.bin_serialise_into(& mut self.forward_digest)?;

        let clone = self.forward_digest.clone();

        let digest = clone.finalize();

        relay.set_digest(u32::from_be_bytes(digest[0..4].try_into().unwrap()));

        Ok(())
    }

    pub fn verify_backward_digest(& mut self, relay: & mut RelayCell) -> torserde::Result<()> {

        let sent_digest = relay.get_digest();

        relay.set_digest(0);

        relay.bin_serialise_into(& mut self.backward_digest)?;

        let clone = self.backward_digest.clone();

        let calculated_digest = u32::from_be_bytes((&clone.finalize()[0..4]).try_into().unwrap());

        println!("calculated: {}", calculated_digest);

        if sent_digest != calculated_digest {
            return Err(torserde::ErrorKind::BadDigest(sent_digest, calculated_digest));
        }

        Ok(())
    }

    pub fn encrypt(& mut self, relay: RelayCell) -> torserde::Result<Encrypted> {

        let mut array = [0u8; 509];

        relay.bin_serialise_into(array.as_mut())?;

        self.forward_encryptor.apply_keystream(array.as_mut());

        Ok(Encrypted(array))

    }

    pub fn decrypt(& mut self, relay: & Encrypted) -> torserde::Result<RelayCell> {

        let mut array = relay.0;

        self.backward_decryptor.apply_keystream(array.as_mut());

        RelayCell::bin_deserialise_from(array.as_ref())
    }
}

src/cells.rs

use std::net::{IpAddr, Ipv4Addr};
use chrono::{DateTime, Local};

use torserde_macros::Torserde;
use torserde::{TorSerde, NLengthVector, VersionsVector};
use std::io::{Read, Write};
use std::borrow::{BorrowMut};

use lazy_static::lazy_static;
use ring::rand::SecureRandom;

use crate::misc::UnpackedCell;

lazy_static!{
    static ref CSRNG: ring::rand::SystemRandom = ring::rand::SystemRandom::new();
}

#[derive(Debug, Torserde)]
pub struct Cert {
    cert_type: u8,
    certificate: NLengthVector<u8, 2>,
}

#[derive(Debug, Torserde)]
pub struct LinkSpecifier {
    ltype: u8,
    lspec: NLengthVector<u8, 1>
}

#[derive(Debug, Torserde)]
#[repr(u8)]
pub enum SendMePayload {
    Ignore = 0,
    Authenticated { length: u16, digest: [u8; 20] } = 1,
}

//Represents the payload of an encrypted cell
#[derive(Debug, Torserde)]
pub struct Encrypted(pub [u8; 509]);

#[derive(Debug, Torserde)]
#[repr(u8)]
pub enum EndReason {
    Misc = 1,
    ResolveFailed = 2,
    ConnectRefused = 3,
    ExitPolicy{ ip: Ipv4Addr, ttl: u32 } = 4,
    Destroy = 5,
    Done = 6,
    Timeout = 7,
    NoRoute = 8,
    Hibernating = 9,
    Internal = 10,
    ResourceLimit = 11,
    ConnReset = 12,
    TorProtocol = 13,
    NotDirectory = 14,
}

#[derive(Debug, Torserde)]
#[repr(u8)]
pub enum DestroyReason {
    None = 0,
    Protocol = 1,
    Internal = 2,
    Requested = 3,
    Hibernating = 4,
    ResourceLimit = 5,
    ConnectFailed = 6,
    OrIdentity = 7,
    OrConnClosed = 8,
    Finished = 9,
    Timeout = 10,
    Destroyed = 11,
    NoSuchService = 12,

}

#[derive(Debug, Torserde)]
#[repr(u8)]
pub enum Relay {
    Begin{ addr_and_port: String, flags: u32 } = 1, //Done
    Data{ data: [u8; 498] } = 2, //DOne - Be careful with this as moving [u8; 498] is expensive
    End{ end_reason: EndReason } = 3, //Done
    Connected { ip: Ipv4Addr, ttl: u32 } = 4, //Done
    SendMe { payload: SendMePayload } = 5, //Done

    Truncated{ reason: DestroyReason } = 9, //Done

    BeginDir = 13,
    Extend2{ link_specifiers: NLengthVector<LinkSpecifier, 1>, htype: u16, handshake_data: NLengthVector<u8, 2> } = 14, //What is a link specifier?
    Extended2{ handshake_data: NLengthVector<u8, 2> } = 15, //Done
    EstablishRendezvous{ rendezvous_cookie: [u8; 20] } = 33, //Done
    Introduce1 = 34, //Need to do this one
    // Rendezvous2{ handshake_data: & 'a [u8] } = 37, // Length?
    RendezvousEstablished = 39,
    IntroduceAck = 40,
}

#[derive(Debug, Torserde)]
#[repr(u8)]
pub enum Command {
    /* Fixed length commands */
    Padding = 0,
    Create{ onion_skin: [u8; 20] } = 1, //Done
    //Created = 2,
    Relay{ contents: Encrypted } = 3, //Figure out padding and encryption
    Destroy{ reason: DestroyReason } = 4, //Done
    CreateFast{ onion_skin: [u8; 20] } = 5, //Done
    CreatedFast{ handshake_data: [u8; 40] } = 6, //Done
    NetInfo{ timestamp: DateTime<Local>, other_ip: IpAddr, this_ips: NLengthVector<IpAddr, 1> } = 8, //Done
    RelayEarly{ contents: Encrypted } = 9, //Figure out padding and encryption
    Create2{ handshake_type: u16, onion_skin: NLengthVector<u8, 2> } = 10, //Done
    Created2{ handshake_data: NLengthVector<u8, 2> } = 11, //Done

    /* Variable length commands */
    Versions{ version_list: VersionsVector } = 7, //Done
    //VPadding = 128,
    Certs{ length: u16, certs: NLengthVector<Cert, 1> } = 129, //Done
    AuthChallenge{ length: u16, challenge: [u8; 32], methods: NLengthVector<u16, 2> } = 130, //Done
    //Authenticate = 131,
    //Authorize = 132,
}

#[derive(Debug)]
pub struct RelayCell {
    command: u8,
    recognised: u16,
    stream_id: u16,
    digest: u32,
    data: NLengthVector<u8, 2>,
    padding: Option<Vec<u8>>,


}

impl TorSerde for RelayCell {
    fn bin_serialise_into<W: Write>(&self, mut stream: W) -> torserde::Result<u32> {
        let command_length = self.command.bin_serialise_into(stream.borrow_mut())?;

        let recognised_length = self.recognised.bin_serialise_into(stream.borrow_mut())? ;

        let stream_id_length = self.stream_id.bin_serialise_into(stream.borrow_mut())? ;

        let digest_length = self.digest.bin_serialise_into(stream.borrow_mut())? ;

        let data_length = self.data.bin_serialise_into(stream.borrow_mut())? ;

        let padding_length = std::io::copy(& mut (&self.padding.as_ref().unwrap()[..]), stream.borrow_mut())? as u32;

        let serialised_length = command_length + recognised_length + stream_id_length + digest_length + data_length + padding_length;

        if serialised_length != 509 {
            return Err(torserde::ErrorKind::InvalidRelayLength(serialised_length, command_length, recognised_length, stream_id_length, digest_length, data_length, padding_length))
        }

        Ok(509)
    }

    fn bin_deserialise_from<R: Read>(mut stream: R) -> torserde::Result<Self> {
        let command = u8::bin_deserialise_from(stream.borrow_mut())?;
        let recognised = u16::bin_deserialise_from(stream.borrow_mut())?;
        let stream_id = u16::bin_deserialise_from(stream.borrow_mut())?;
        let digest = u32::bin_deserialise_from(stream.borrow_mut())?;
        let data = <NLengthVector<u8, 2>>::bin_deserialise_from(stream.borrow_mut())?;

        let expected_padding_length = 509 - 11 - data.0.len();

        let mut taken = stream.borrow_mut().take(expected_padding_length as u64);

        let mut padding = Vec::new();

        let read_padding_length = taken.read_to_end(& mut padding)?;

        if read_padding_length != expected_padding_length {
            return Err(torserde::ErrorKind::NotEnoughPadding(expected_padding_length, read_padding_length));
        }

        let padding = Some(padding);

        Ok(Self {
            command,
            recognised,
            stream_id,
            digest,
            data,
            padding,
        })
    }

    fn serialised_length(&self) -> u32 {
        509
    }
}

impl RelayCell {

    pub fn new(stream_id: u16, contents: Relay) -> Self {
        let recognised = 0;
        let digest = 0;
        let unpacked = Self::get_vector(contents).unwrap();

        let mut padding: Vec<_> = (0..509-11-unpacked.data_length()).into_iter().map(|_| 0u8).collect();

        CSRNG.fill(& mut padding).unwrap();

        let padding = Some(padding);

        Self {
            command: unpacked.command(),
            recognised,
            stream_id,
            digest,
            data: NLengthVector::<u8, 2>::from(unpacked.data()),
            padding,
        }
    }

    pub fn set_digest(& mut self, digest: u32) {
        self.digest = digest;
    }

    pub fn get_digest(& self) -> u32 {
        self.digest
    }

    fn get_vector(relay: Relay) -> torserde::Result<UnpackedCell> {
        let mut unpacked = UnpackedCell::default();

        relay.bin_serialise_into(& mut unpacked)?;

        Ok(unpacked)
    }

    pub fn get_payload(self) -> torserde::Result<Option<Relay>> {

        if self.data.0.is_empty() {
            Ok(None)
        } else {
            let unpacked = UnpackedCell::new(self.command, Some(self.data.0));

            Ok(Some(Relay::bin_deserialise_from(unpacked)?))
        }

    }



}

impl Command {

    fn get_var_len(&self) -> Option<u32> {

        match &self {
            Command::Versions { version_list } => { Some((version_list.0.len()*2) as u32) }
            Command::Certs { length, certs: _ } => { Some(*length as u32) }
            Command::AuthChallenge { length, challenge: _, methods: _ } => { Some(*length as u32) }
            _ => { None }
        }

    }

    fn is_var_len(&self) -> bool {
        self.get_var_len().is_some()
    }

    ///If the command is not recognised (and not serialisable into `Command`) then we get the length of the cell so we can discard it
    fn is_var_command(command: u8) -> bool {
        command == 7 || command >= 128
    }


}

#[derive(Debug)]
pub struct TorCell {
    circuit_id: u32,
    payload: Command, //Contains the command, length (optional) and payload
}

impl<'a> TorCell {
    pub fn new(circuit_id: u32, payload: Command) -> Self {

        Self {
            circuit_id,
            payload,
        }
    }

    pub fn get_command(& self) -> &Command {
        &self.payload
    }

    ///Keep trying `from_stream` until the result is not a Err(DiscardedCell)
    pub fn try_from_stream<R: Read>(mut stream: R, version: u32) -> torserde::Result<Self> {
        todo!()
    }

    pub fn from_stream<R: Read>(mut stream: R, version: u32) -> torserde::Result<Self> {
        let circuit_id = if version < 4 {
            u16::bin_deserialise_from(stream.borrow_mut())? as u32
        } else {
            u32::bin_deserialise_from(stream.borrow_mut())?
        };

        //Deserialise the command, then use that to get the rest of the cell into a buffer, then deserialise the buffer
        //This means that even if a deserialisation stops mid way, we still have removed the cell from the stream

        match Command::bin_deserialise_from(stream.borrow_mut()) {
            Ok(payload) => {
                let padding_length =  if let Some(length) = payload.get_var_len() {
                    if length > payload.serialised_length() - 1 {
                        length - (payload.serialised_length() - 1)
                    } else {
                        0
                    }
                } else {
                    509 - (payload.serialised_length() - 1)
                };

                //Discard the remaining padding bytes
                if padding_length != 0 {
                    let mut taken = stream.borrow_mut().take(padding_length as u64);
                    println!("     padding: {}", padding_length);
                    std::io::copy(&mut taken, &mut crate::misc::NullStream)?;
                }

                Ok(Self {
                    circuit_id,
                    payload
                })
            },
            Err(kind) => match kind {
                torserde::ErrorKind::BadDiscriminant(discriminant) => {

                    let bytes_left_in_cell = if Command::is_var_command(discriminant as u8) {
                        todo!()
                        //Serialise the next few bytes to get the length
                    } else {
                        509
                    };

                    //Discard `bytes_left_in_cell` bytes from stream



                    Err(torserde::ErrorKind::DiscardedCell(discriminant))
                },
                _ => Err(kind)
            }
        }



    }

    pub fn into_stream<W: Write>(self, mut stream: W, version: u32) -> torserde::Result<()> {
        if version < 4 {
            (self.circuit_id as u16).bin_serialise_into(stream.borrow_mut())?;
        } else {
            (self.circuit_id as u32).bin_serialise_into(stream.borrow_mut())?;
        }

        self.payload.bin_serialise_into(stream.borrow_mut())?; //Subtract one since out payload includes the command byte

        if !self.payload.is_var_len() {
            for _ in 0..(510-self.payload.serialised_length()) {
                0u8.bin_serialise_into(stream.borrow_mut())?;
            }
        }

        Ok(())

    }

}